std/range/package.d

```diff
@@ -9594,6 +9594,324 @@ auto refRange(R)(R* range)
     foo(r);
 }
 
+/**
+Bitwise adapter over Integral type Ranges. Consumes the range elements bit by bit.
+*/
+struct Bitwise(R)
+    if (isInputRange!R && isIntegral!(ElementType!R))
+{
+    alias ER = ElementType!R;
+    alias UER = Unsigned!ER;
+
+    private R parent;
+    private UER mask;
+    private int maskPos;
+    private ER accumulator;
+
+    private enum UER bitsNum = ER.sizeof * 8;
+
+    this()(auto ref R range)
+    {
+        parent = range;
+
+        static if (hasLength!R)
+        {
+            len = parent.length * bitsNum;
+        }
+
+        initPrivates();
+
+        static if (isBidirectionalRange!R)
+        {
+            initBackPrivates();
+        }
+    }
+
+    bool empty() const
+    {
+        static if (hasLength!R)
+        {
+            return len == 0;
+        }
+        else
+        {
+            /*
+               If we consumed the last element of the range, but not all the
+               bits in the last element
+             */
+            return parent.empty && (maskPos == 0);
+        }
+    }
+
+    bool front()
+    {
+        return cast(bool)(accumulator & mask);
+    }
+
+    void popFront()
+    {
+        if (mask == 0)
+        {
+            initPrivates();
+        }
+        else
+        {
+            mask >>>= 1;
+            --maskPos;
```

Comment:
So I see mask and maskPos are redundant (can be deduced from one another). No need to maintain both. I suggest you write a method turning one into the other. The cheapest way is:

```
auto mask() { return (1U << maskPos) - 1U; }
```

Then just use `mask` as up until now. (I may be off by 1 or something so please double check.)